probert/filesystem.py

```diff
@@ -22,23 +24,142 @@
 log = logging.getLogger('probert.filesystems')
 
 
-def get_device_filesystem(device):
+def run(cmdarr, **kw):
+    try:
+        return subprocess.check_output(cmdarr, universal_newlines=True, **kw)
+    except subprocess.CalledProcessError as cpe:
+        if cpe.stderr:
+            log.debug('stderr: %s', cpe.stderr)
+        log.exception(cpe)
+        return None
+
+
+def get_dumpe2fs_info(path):
+    ret = {}
+    out = run(['dumpe2fs', '-h', path])
+    if out is None:
+        log.debug('ext volume size not found: dumpe2fs failure')
+        return None
+    # Block count:              20480
+    # Block size:               4096
+    block_count_matcher = re.compile(r'^Block count:\s+(\d+)$')
+    block_size_matcher = re.compile(r'^Block size:\s+(\d+)$')
+    for line in out.splitlines():
+        m = block_count_matcher.fullmatch(line)
+        if m:
+            ret['block_count'] = int(m.group(1))
+        m = block_size_matcher.fullmatch(line)
+        if m:
+            ret['block_size'] = int(m.group(1))
+    if 'block_count' not in ret or 'block_size' not in ret:
+        log.debug('ext volume size not found: unexpected output format')
+        return None
+    return ret
+
+
+def get_resize2fs_info(path):
+    # Estimated minimum size of the filesystem: 1696
+    out = run(['resize2fs', '-P', path])
+    if out is None:
+        return None
+    min_blocks_matcher = re.compile(
+        r'^Estimated minimum size of the filesystem: (\d+)$')
+    for line in out.splitlines():
+        m = min_blocks_matcher.fullmatch(line)
+        if m:
+            return {'min_blocks': int(m.group(1))}
+    return None
+
+
+def get_ext_sizing(device):
+    path = device.device_node
+    dumpe2fs_info = get_dumpe2fs_info(path)
+    if not dumpe2fs_info:
+        return None
+    ret = {'SIZE': dumpe2fs_info['block_count'] * dumpe2fs_info['block_size']}
+
+    resize2fs_info = get_resize2fs_info(path)
+    if resize2fs_info:
+        min_size = resize2fs_info['min_blocks'] * dumpe2fs_info['block_size']
+        ret['ESTIMATED_MIN_SIZE'] = min_size
+    return ret
```

Comment:
question: if resize2fs_info is None, ESTIMATED_MIN_SIZE would not be in the ret dict. Is this something that the caller handles?

Reply:
Only in some cases.  I have expanded this.